03 Feature Engineering

1. Objective
2. Business Question

3. Load Dataset

4. Feature Selection
    4.1 Remove Unnecessary Features
    4.2 Feature Categorization

5. Missing Value Handling
    5.1 Analyze Missing Strategy
    5.2 Numerical Features
    5.3 Categorical Features

6. Feature Transformation
    6.1 Transaction Features
    6.2 Card Features
    6.3 Identity Features
    6.4 Time Features

7. Feature Encoding

8. Final Feature Set

9. Save Engineered Dataset

In [1]:
import pandas as pd
import numpy as np
from pathlib import Path

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

# Project paths (absolute to avoid working directory issues)
PROJECT_DIR = Path('/home/rizalkurnia/Projects/DataAnalysis/fraud_detection')
DATA_RAW = PROJECT_DIR / 'data' / 'raw'
OUTPUT_FIG = PROJECT_DIR / 'outputs' / 'figures'
DATA_PROCESSED = PROJECT_DIR / 'data' / 'processed'

# Feature Engineering

Preparing Machine Learning-Ready Features for the Fraud Risk Decision Engine

# 01 Introduction

## Objective

Transform the merged transaction and identity dataset into a machine learning-ready feature set through feature selection, missing value handling, feature transformation, and feature encoding.

## Scope

This notebook focuses on:

- Feature selection
- Missing value handling
- Feature transformation
- Feature encoding
- Feature scaling (if required)
- Final feature set preparation


# 02 Dataset Preparation

## Objective

Load and verify the merged dataset generated during the exploratory data analysis stage to ensure it is ready for feature engineering.

## Business Question

**Is the dataset complete and ready for feature engineering?**


### 2.1 Load Dataset

The merged dataset generated during the exploratory data analysis stage is loaded as the input for the feature engineering process.

In [2]:
df = pd.read_pickle(
    DATA_PROCESSED / "merged_dataset.pkl"
)

In [3]:
print(df.shape)
print(df["isFraud"].value_counts())
df.head()

(590540, 435)
isFraud
0    569877
1     20663
Name: count, dtype: int64


,TransactionID,isFraud,TransactionDT,TransactionAmt,ProductCD,card1,card2,card3,card4,card5,card6,addr1,addr2,dist1,dist2,P_emaildomain,R_emaildomain,C1,C2,C3,C4,C5,C6,C7,C8,C9,C10,C11,C12,C13,C14,D1,D2,D3,D4,D5,D6,D7,D8,D9,D10,D11,D12,D13,D14,D15,M1,M2,M3,M4,M5,M6,M7,M8,M9,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,V11,V12,V13,V14,V15,V16,V17,V18,V19,V20,V21,V22,V23,V24,V25,V26,V27,V28,V29,V30,V31,V32,V33,V34,V35,V36,V37,V38,V39,V40,V41,V42,V43,V44,V45,V46,V47,V48,V49,V50,V51,V52,V53,V54,V55,V56,V57,V58,V59,V60,V61,V62,V63,V64,V65,V66,V67,V68,V69,V70,V71,V72,V73,V74,V75,V76,V77,V78,V79,V80,V81,V82,V83,V84,V85,V86,V87,V88,V89,V90,V91,V92,V93,V94,V95,V96,V97,V98,V99,V100,V101,V102,V103,V104,V105,V106,V107,V108,V109,V110,V111,V112,V113,V114,V115,V116,V117,V118,V119,V120,V121,V122,V123,V124,V125,V126,V127,V128,V129,V130,V131,V132,V133,V134,V135,V136,V137,V138,V139,V140,V141,V142,V143,V144,V145,V146,V147,V148,V149,V150,V151,V152,V153,V154,V155,V156,V157,V158,V159,V160,V161,V162,V163,V164,V165,V166,V167,V168,V169,V170,V171,V172,V173,V174,V175,V176,V177,V178,V179,V180,V181,V182,V183,V184,V185,V186,V187,V188,V189,V190,V191,V192,V193,V194,V195,V196,V197,V198,V199,V200,V201,V202,V203,V204,V205,V206,V207,V208,V209,V210,V211,V212,V213,V214,V215,V216,V217,V218,V219,V220,V221,V222,V223,V224,V225,V226,V227,V228,V229,V230,V231,V232,V233,V234,V235,V236,V237,V238,V239,V240,V241,V242,V243,V244,V245,V246,V247,V248,V249,V250,V251,V252,V253,V254,V255,V256,V257,V258,V259,V260,V261,V262,V263,V264,V265,V266,V267,V268,V269,V270,V271,V272,V273,V274,V275,V276,V277,V278,V279,V280,V281,V282,V283,V284,V285,V286,V287,V288,V289,V290,V291,V292,V293,V294,V295,V296,V297,V298,V299,V300,V301,V302,V303,V304,V305,V306,V307,V308,V309,V310,V311,V312,V313,V314,V315,V316,V317,V318,V319,V320,V321,V322,V323,V324,V325,V326,V327,V328,V329,V330,V331,V332,V333,V334,V335,V336,V337,V338,V339,card5_group,id_01,id_02,id_03,id_04,id_05,id_06,id_07,id_08,id_09,id_10,id_11,id_12,id_13,id_14,id_15,id_16,id_17,id_18,id_19,id_20,id_21,id_22,id_23,id_24,id_25,id_26,id_27,id_28,id_29,id_30,id_31,id_32,id_33,id_34,id_35,id_36,id_37,id_38,DeviceType,DeviceInfo
0,2987000,0,86400,68.5,W,13926,NaN,150.0,discover,142.0,credit,315.0,87.0,19.0,NaN,NaN,NaN,1.0,1.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,2.0,0.0,1.0,1.0,14.0,NaN,13.0,NaN,NaN,NaN,NaN,NaN,NaN,13.0,13.0,NaN,NaN,NaN,0.0,T,T,T,M2,F,T,NaN,NaN,NaN,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,0.0,0.0,1.0,1.0,1.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,1.0,1.0,1.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.0,1.0,1.0,1.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,1.0,1.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0,1.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,0.0,117.0,0.0,0.0,0.0,0.0,0.0,117.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,0.0,0.0,1.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,117.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,117.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Other,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2987001,0,86401,29.0,W,2755,404.0,150.0,mastercard,102.0,credit,325.0,87.0,NaN,NaN,gmail.com,NaN,1.0,1.0,0.0

### Observation

The merged dataset was successfully loaded and contains **590,540 transactions** with **434 features**, including the target variable (`isFraud`).

The target distribution remains unchanged from the exploratory data analysis stage, indicating that no records were lost during data persistence. This confirms that the dataset is ready for feature engineering.

# 03 Feature Selection

## Objective

Identify which features should be retained for subsequent preprocessing and machine learning modeling while removing variables that do not provide predictive information.

## Business Question

**Which features should be retained for feature engineering and fraud prediction?**

### 3.1 Remove Unnecessary Features

The merged dataset contains one identifier variable, `TransactionID`, which uniquely identifies each transaction but does not contribute predictive information for fraud detection. Therefore, this feature is removed before proceeding with feature engineering.

In [4]:
# Remove identifier column if it still exists

if "TransactionID" in df.columns:
    df = df.drop(columns=["TransactionID"])

print(df.shape)

(590540, 434)


### 3.2 Feature Categorization

The remaining variables are grouped into numerical and categorical features based on their data types. This categorization provides a structured foundation for subsequent preprocessing tasks, including missing value handling, feature transformation, feature encoding, and model development.

In [5]:
# Target variable
target_col = "isFraud"

# Numerical features
numerical_features = (
    df.select_dtypes(include=["number"])
      .columns
      .drop(target_col)
      .tolist()
)

# Categorical features
categorical_features = (
    df.select_dtypes(include=["object", "string", "category"])
      .columns
      .tolist()
)

print(f"Numerical Features   : {len(numerical_features)}")
print(f"Categorical Features : {len(categorical_features)}")
print(f"Target               : {target_col}")

Numerical Features   : 401
Categorical Features : 32
Target               : isFraud


### Observation

After removing the identifier column, the dataset contains **401 numerical features**, **32 categorical features**, and **1 target variable (`isFraud`)**.

This categorization establishes the foundation for subsequent preprocessing steps, allowing different feature types to be handled using appropriate strategies during missing value treatment, feature transformation, and encoding.

# 04 Missing Value Handling

## Objective

Identify and handle missing values using strategies that preserve meaningful information while minimizing the introduction of bias into the dataset.

## Business Question

**How should missing values be treated to produce a reliable feature set for fraud prediction?**

### 4.1 Missing Value Overview

The exploratory data analysis revealed that missing values are widely distributed across the dataset. Some variables contain only a small proportion of missing observations, whereas others—particularly several identity-related attributes—contain missing values in more than 90% of the records.

Rather than removing all highly incomplete features, the missing-value treatment strategy will be determined based on both the amount of missing data and the business meaning of each feature.

In [6]:
missing_summary = (
    df.isnull()
      .sum()
      .to_frame("Missing")
)

missing_summary["Missing (%)"] = (
    missing_summary["Missing"] / len(df) * 100
).round(2)

missing_summary = (
    missing_summary
    .query("Missing > 0")
    .sort_values("Missing (%)", ascending=False)
)

missing_summary.head(20)

,Missing,Missing (%)
id_24,585793,99.20
id_07,585385,99.13
id_25,585408,99.13
id_08,585385,99.13
id_26,585377,99.13
id_21,585381,99.13
id_23,585371,99.12
id_22,585371,99.12
id_27,585371,99.12
dist2,552913,93.63


### 4.2 Missing Value Strategy

Missing values in this dataset originate from different sources and therefore cannot be handled using a single imputation strategy. Based on the exploratory data analysis, missing values are categorized into three groups:

1. **Ordinary transaction features**, where missing values are assumed to result from incomplete observations and can be imputed using statistical methods.

2. **Identity-related features**, where missing values often indicate that no identity information was collected for the transaction. In this case, the absence of information itself may carry predictive value and should therefore be preserved whenever appropriate.

3. **Extremely sparse features**, which contain a very high proportion of missing values. These features will be evaluated individually to determine whether they should be retained, transformed, or removed, depending on both their business meaning and predictive potential.

The following sections apply these principles to construct a consistent and machine learning-ready dataset.

### 4.3 Apply Missing Value Treatment

The missing value treatment is performed according to the strategy defined in the previous section. Different feature groups are handled using different approaches based on their business meaning and data characteristics.

The treatment process follows these principles:

- Numerical transaction features are imputed using statistical values.
- Categorical features preserve missing information as an explicit category.
- Identity-related features retain missingness whenever it represents meaningful behavioral information.
- Extremely sparse features are evaluated individually before deciding whether to retain or remove them.

#### 4.3.1 Missing Value Treatment Rules

The following rules are applied during missing value treatment:

| Feature Group | Treatment | Rationale |
|---------------|-----------|-----------|
| Transaction numerical features | Median imputation | Preserves the overall distribution while reducing the influence of extreme values. |
| Transaction categorical features | `"Missing"` category | Missing values are treated as a valid transaction category. |
| Identity numerical features | `-999` | Missing identity information may represent meaningful transaction behavior rather than incomplete data. A sentinel value preserves this information for subsequent modeling. |
| Identity categorical features | `"Missing"` category | Preserves the absence of identity information as a potential predictive signal. |
| Extremely sparse features | Evaluated individually | Features with extremely high missing rates are assessed separately before deciding whether to retain or remove them. |

#### 4.3.2 Missing Value Imputation

The missing value treatment rules established in the previous section are applied to each feature group.

To ensure consistency with the exploratory data analysis, different imputation methods are used according to the characteristics of each feature group. Transaction-related numerical features are imputed using statistical values, whereas identity-related features preserve missingness as meaningful behavioral information through sentinel values or explicit categories.

The following implementation applies these rules across the dataset.

In [19]:
# ===============================
# Identify feature groups
# ===============================

identity_columns = [
    col for col in df.columns
    if col.startswith("id_")
]

identity_numeric = [
    col for col in identity_columns
    if col in numerical_features
]

identity_categorical = [
    col for col in identity_columns
    if col in categorical_features
]

transaction_numeric = [
    col for col in numerical_features
    if col not in identity_numeric
]

transaction_categorical = [
    col for col in categorical_features
    if col not in identity_categorical
]

print(f"Transaction numerical features   : {len(transaction_numeric)}")
print(f"Transaction categorical features : {len(transaction_categorical)}")
print(f"Identity numerical features      : {len(identity_numeric)}")
print(f"Identity categorical features    : {len(identity_categorical)}")


# ===============================
# Apply imputation
# ===============================

# Transaction numerical features
for col in transaction_numeric:
    if df[col].isna().any():
        df[col] = df[col].fillna(df[col].median())

# Identity numerical features
for col in identity_numeric:
    df[col] = df[col].fillna(-999)

# Transaction categorical features
for col in transaction_categorical:
    df[col] = df[col].fillna("Missing")

# Identity categorical features
for col in identity_categorical:
    df[col] = df[col].fillna("Missing")

Transaction numerical features   : 378
Transaction categorical features : 17
Identity numerical features      : 23
Identity categorical features    : 15


The feature groups are separated before applying imputation because transaction and identity features follow different missing value strategies.

Transaction features primarily represent transaction behavior and are treated using conventional statistical imputation, whereas identity features preserve missingness as a potentially informative behavioral signal identified during the exploratory data analysis.

#### 4.3.3 Verification

After applying the missing value treatment, the dataset is verified to ensure that the predefined strategy has been implemented successfully.

The verification focuses on confirming that:

- Missing values have been handled according to their respective feature groups.
- No unexpected missing values remain in the dataset.
- The dataset is ready for subsequent feature engineering and model development.

In [20]:
remaining_missing = (
    df.isna()
      .sum()
      .sort_values(ascending=False)
)

remaining_missing = remaining_missing[remaining_missing > 0]

print(f"Remaining features with missing values : {len(remaining_missing)}")

remaining_missing.head(20)

Remaining features with missing values : 0


Series([], dtype: int64)

In [21]:
print(f"Dataset shape : {df.shape}")
print(f"Total missing values : {df.isna().sum().sum()}")

Dataset shape : (590540, 434)
Total missing values : 0


The verification confirms that all missing values have been treated according to the predefined strategy. No remaining missing values are found in the dataset, indicating that the feature engineering process has successfully produced a complete dataset for the next stage of model development.

# 05 Save Feature Engineered Dataset

## Objective

Store the feature-engineered dataset after feature selection and missing value treatment so it can be reused during the model development stage without repeating the preprocessing pipeline.

## Business Question

**Is the dataset ready to be used for fraud prediction model development?**

In [22]:
# Save feature engineered dataset

df.to_pickle(DATA_PROCESSED / "feature_engineered.pkl")

print("Feature engineered dataset saved successfully.")
print(f"Location : {DATA_PROCESSED / 'feature_engineered.pkl'}")
print(f"Shape    : {df.shape}")

Feature engineered dataset saved successfully.
Location : /home/rizalkurnia/Projects/DataAnalysis/fraud_detection/data/processed/feature_engineered.pkl
Shape    : (590540, 434)


## Summary

This notebook completed the feature engineering stage by:

- Selecting relevant transaction and identity features.
- Categorizing features according to their data types.
- Applying a feature-specific missing value treatment strategy.
- Producing a complete machine learning-ready dataset without remaining missing values.

The resulting dataset will serve as the input for the **Model Development** stage, where the fraud prediction models will be trained and evaluated.